lib/core/src/sync/mod.rs

```diff
@@ -0,0 +1,350 @@
+pub(crate) mod client;
+pub(crate) mod model;
+
+use std::sync::Arc;
+
+use anyhow::{anyhow, Result};
+use futures_util::TryFutureExt;
+use log::{debug, warn};
+use std::collections::HashMap;
+use tokio::sync::{watch, Mutex};
+use tokio_stream::StreamExt as _;
+
+use self::client::SyncerClient;
+use self::model::sync::TrackChangesRequest;
+use self::model::{
+    sync::{ListChangesRequest, Record, SetRecordReply, SetRecordRequest, SetRecordStatus},
+    DecryptedRecord, SyncData,
+};
+use self::model::{Merge, OutboundChange, SyncDetails};
+use crate::prelude::{ChainSwap, ReceiveSwap, SendSwap};
+use crate::{model::Signer, persist::Persister, utils};
+
+const CURRENT_SCHEMA_VERSION: f32 = 0.01;
+const MAX_SET_RECORD_REATTEMPTS: u8 = 5;
+
+pub(crate) struct SyncService {
+    connect_url: String,
+    persister: Arc<Persister>,
+    signer: Arc<Box<dyn Signer>>,
+    client: Box<dyn SyncerClient>,
+    outbound_changes: Mutex<HashMap<String, OutboundChange>>,
```

Comment:
Not sure we need the outbound_changes, we can have that info in the db (additional field that saves the changed fields array).
Also not sure we need a max_attempt mechanism here, I would try forever for each item that fails. It will propagate unexpected bugs to the surface and will allow us to fix them.